### PR TITLE
Allow cross-group subresource registrations in the APIInstaller.

### PR DESCRIPTION
**Pull request by madhusudancs**
@lavalamp @bgrant0607 @deads2k @smarterclayton @liggitt @piosz 

This is a rework of #20745 and #19802 based on all your reviews. Only the last commit needs review here. The other commits are being reviewed separately in #21603.

### CODE REVIEW

File: pkg/apiserver/api_installer.go [hunk 1]
```diff
@@ -110,62 +110,54 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
...
+	if override, ok := a.group.SubresourceOverrides[path]; ok {
+		allowedGroup = override.GroupVersion.Group
 	}
-	hasSubresource := len(subresource) > 0
 
 	object := storage.New()
 	fqKinds, err := a.group.Typer.ObjectKinds(object)
 	if err != nil {
-		return nil, err
+		return fqKindToRegister, err
```
Comment (deads2k) [inline pkg/apiserver/api_installer.go]: This is just an empty `unversioned.GroupVersionKind`?  Please return that explictly to make it clear the value is meant to be invalid/unused.

File: pkg/apiserver/api_installer.go [hunk 2]
```diff
@@ -110,62 +110,54 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
...
-		// TODO: support deeper paths
-		return nil, fmt.Errorf("api_installer allows only one or two segment paths (resource or resource/subresource)")
+// validFQKind validates and returns the group version kind registered for the given storage
+// object. If the storage object is a subresource and has an override specified, the object's
+// group version is validated against the group version in the override for that subresource.
+func (a *APIInstaller) validFQKind(path string, storage rest.Storage) (unversioned.GroupVersionKind, error) {
+	fqKindToRegister := unversioned.GroupVersionKind{}
+	allowedGroup := a.group.GroupVersion.Group
+	if override, ok := a.group.SubresourceOverrides[path]; ok {
+		allowedGroup = override.GroupVersion.Group
```
Comment (deads2k) [inline pkg/apiserver/api_installer.go]: So the `SubresourceOverrides` explicitly provides a `GroupVersion`, but we're ignoring the `version` part of that information?  Why?

File: pkg/apiserver/api_installer.go [hunk 3]
```diff
@@ -110,62 +110,54 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
...
 	// a given go type can have multiple potential fully qualified kinds.  Find the one that corresponds with the group
 	// we're trying to register here
-	fqKindToRegister := unversioned.GroupVersionKind{}
 	for _, fqKind := range fqKinds {
-		if fqKind.Group == a.group.GroupVersion.Group {
-			fqKindToRegister = fqKind
+		// DEBUG: if fqKind.Group == a.group.GroupVersion.Group
+		if fqKind.Group == allowedGroup {
+			fqKindToRegister.Group = fqKind.Group
+			fqKindToRegister.Version = a.group.GroupVersion.Version
```
Comment (deads2k) [inline pkg/apiserver/api_installer.go]: This doesn't look right.  Why would I take group information from one source and smash it together with version information from another source which is referring to a potentially unrelated `GroupVersion`?  I could probably see it if `allowedGroup == a.group.GroupVersion.Group` (in which case this entire loop in unnecessary), but why do this otherwise?

File: pkg/apiserver/api_installer.go [hunk 4]
```diff
@@ -110,62 +110,54 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
...
+			fqKindToRegister.Kind = fqKind.Kind
 			break
 		}
 
 		// TODO This keeps it doing what it was doing before, but it doesn't feel right.
 		if fqKind.Group == extensions.GroupName && fqKind.Kind == "ThirdPartyResourceData" {
```
Comment (deads2k) [inline pkg/apiserver/api_installer.go]: Do you know why this is special-cased?  I still don't, but this pull is changing behavior around this bit of hardcoding (unlike the previous comment) so I think it should be re-visited.

File: pkg/apiserver/api_installer.go (hunk 2 above)
Comment (deads2k) [inline pkg/apiserver/api_installer.go]: Yeah, since version is definitely expected in the return and you're using this field as the marker, why is only half the field respected?  It almost seems like what you want is to take the `GroupVersion` from here and do a `.WithKind` from the `ObjectKinds`, which matches the expected group.

File: pkg/apiserver/api_installer.go [hunk 5]
```diff
@@ -948,3 +973,19 @@ var _ rest.StorageMetadata = defaultStorageMetadata{}
...
+	switch parts := strings.Split(path, "/"); len(parts) {
+	case 2:
+		resource, subresource = parts[0], parts[1]
+	case 1:
+		resource = parts[0]
+	default:
+		// TODO: support deeper paths
+		return "", "", false, fmt.Errorf("api_installer allows only one or two segment paths (resource or resource/subresource)")
+	}
+	return resource, subresource, len(subresource) > 0, nil
```
Comment (deads2k) [inline pkg/apiserver/api_installer.go]: This bool doesn't seem to add much value.  Why not return a single field that can't possibly disagree with itself?

File: pkg/apiserver/api_installer.go (hunk 2 above)
Comment (bgrant0607) [inline pkg/apiserver/api_installer.go]: I had the same question.

File: pkg/apiserver/api_installer.go (hunk 3 above)
Comment (bgrant0607) [inline pkg/apiserver/api_installer.go]: +1

File: pkg/apiserver/api_installer.go (hunk 4 above)
Comment (bgrant0607) [inline pkg/apiserver/api_installer.go]: Let's deal with ThirdPartyResource in another PR. The currently implemented behavior isn't what we want.

File: pkg/apiserver/api_installer.go (hunk 1 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: Done.

File: pkg/apiserver/api_installer.go (hunk 2 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: You are right and I realize this thing is wrong. I should take both group and version from the override. I was trying hard to keep the old code as much as possible. I have fixed this.

File: pkg/apiserver/api_installer.go (hunk 3 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: Fixed this, same as above.

File: pkg/apiserver/api_installer.go (hunk 4 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: Yeah, I think that should be in another PR.

**Comment by madhusudancs** [comment]
@lavalamp I have addressed your comments, PTAL.

@bgrant0607 @deads2k PTAL too.

**Comment by lavalamp** [comment]
Sorry, I didn't get all the way through this yesterday, so I may have more comments. Getting through this is my priority for today.

File: pkg/apiserver/api_installer.go [hunk 6]
```diff
@@ -110,62 +111,55 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
...
-	case 1:
-		resource = parts[0]
-	default:
-		// TODO: support deeper paths
-		return nil, fmt.Errorf("api_installer allows only one or two segment paths (resource or resource/subresource)")
+// validFullyQualifiedKind validates and returns the group version kind registered for the given
+// storage object. If the storage object is a subresource, it just return's the object's group
+// version kind without validating it's group against the group we are installing the REST
+// handlers for here.
+func (a *APIInstaller) validFullyQualifiedKind(path string, storage rest.Storage, hasSubresource bool) (unversioned.GroupVersionKind, error) {
```
Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: Suggest: getBodyKind, getWireKind, getResourceKind

The validation isn't important to highlight in the function name. The thing to call attention to is that this produces the kind corresponding to the return type, not the kind corresponding to the path the resource is mounted under.

File: pkg/apiserver/api_installer.go [hunk 7]
```diff
@@ -187,20 +181,54 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 		parentFQKindToRegister := unversioned.GroupVersionKind{}
```
Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: Can you replace this section with a call to the other function?

File: pkg/apiserver/api_installer.go [hunk 8]
```diff
@@ -187,20 +181,54 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
...
-			return nil, fmt.Errorf("unable to locate fully qualified kind for %v: found %v when registering for %v", reflect.TypeOf(object), fqKinds, a.group.GroupVersion)
+			return nil, fmt.Errorf("unable to locate fully qualified kind for %v: found %v when registering for %v", reflect.TypeOf(parentObject), parentFQKinds, a.group.GroupVersion)
 		}
-
-		parentMapping, err := a.group.Mapper.RESTMapping(parentFQKindToRegister.GroupKind(), a.group.GroupVersion.Version)
+		parentMapping, err := a.group.Mapper.RESTMapping(parentFQKindToRegister.GroupKind(), parentFQKindToRegister.Version)
 		if err != nil {
 			return nil, err
 		}
 		mapping.Scope = parentMapping.Scope
```
Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: Actually, can you just re-call this current function?

File: pkg/apiserver/api_installer.go [hunk 9]
```diff
@@ -110,62 +111,55 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
...
-	versionedPtr, err := a.group.Creater.New(a.group.GroupVersion.WithKind(kind))
-	if err != nil {
-		return nil, err
+// restMapping returns the RESTMapping object for the given resource or the subresource based on
+// the supplied subresource overrides.
+func (a *APIInstaller) restMapping(path, resource string, hasSubresource bool, fqKindToRegister unversioned.GroupVersionKind) (*meta.RESTMapping, error) {
+	mapper := a.group.Mapper
+
+	if hasSubresource {
+		groupMeta, err := registered.Group(fqKindToRegister.Group)
```
Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: Reading this function, I cannot tell if Job's scale subresource will end up exposed at `apis/batch/v1/job/scale` or `apis/autoscaling/v1/job/scale`. The former is correct. Can you add comments until it is clear?

**Comment by lavalamp** [comment]
This needs a test and then I think it's OK.

File: pkg/apiserver/api_installer.go (hunk 9 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: Job will not have a Scale.

Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: Sorry, I keep forgetting. Pretend I said replicaset or deployment.

File: pkg/apiserver/apiserver.go [hunk 10]
```diff
@@ -96,6 +96,11 @@ type APIGroupVersion struct {
 	Context api.RequestContextMapper
 
 	MinRequestTimeout time.Duration
+
+	// SubresourceOverrides contains the GroupVersionKind overrides for each subresource that is
```
Comment (lavalamp) [inline pkg/apiserver/apiserver.go]: Comment doesn't match any more. Mention "external version" somewhere.

File: pkg/genericapiserver/genericapiserver.go [hunk 11]
```diff
@@ -192,6 +192,11 @@ type APIGroupInfo struct {
 	NegotiatedSerializer runtime.NegotiatedSerializer
 	// ParameterCodec performs conversions for query parameters passed to API calls
 	ParameterCodec runtime.ParameterCodec
+
+	// SubresourceOverrides contains the GroupVersionKind overrides for each subresource that is
```
Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: Same comment.

File: pkg/apiserver/apiserver.go [hunk 12]
```diff
@@ -96,6 +96,11 @@ type APIGroupVersion struct {
...
+	// accessible from this API group version. The key of this map should be the path of the
+	// subresource. The keys here should match the keys in the Storage map above for subresources.
+	SubresourceGroupVersionKind map[string]unversioned.GroupVersionKind
```
Comment (lavalamp) [inline pkg/apiserver/apiserver.go]: Why did you need to add it both here and in genericapaiserver? Should only need to be added in one place-- genericapiserver is preferable.

File: pkg/apiserver/api_installer.go [hunk 13]
```diff
@@ -110,96 +110,90 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
...
 	}
+	if fqKindToRegister.IsEmpty() {
+		return unversioned.GroupVersionKind{}, fmt.Errorf("unable to locate fully qualified kind for %v: found %v when registering for %v", reflect.TypeOf(object), fqKinds, a.group.GroupVersion)
+	}
+	return fqKindToRegister, nil
+}
 
-	kind := fqKindToRegister.Kind
+// restMapping returns the RESTMapping object for the given resource or the subresource based on
+// the supplied subresource overrides.
```
Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: Fix comment. "returns rest mapper for the resource, or the resource's parent if it is a sub resource."

File: pkg/apiserver/api_installer.go [hunk 14]
```diff
@@ -110,96 +110,90 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
...
+func (a *APIInstaller) restMapping(path, resource string, hasSubresource bool, fqKindToRegister unversioned.GroupVersionKind) (*meta.RESTMapping, error) {
```
Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: It's extremely strange that this function expects fqKindToRegister to not correspond to resource.

Like maybe just don't take path, hasSubresource or fqKindToRegister as parameters and always do the section in the `if`.

File: pkg/apiserver/api_installer.go (hunk 7 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: Done.

File: pkg/apiserver/api_installer.go (hunk 8 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: Refactored this a bit. So this should be fixed.

**Comment by lavalamp** [comment]
OK, this is looking like it might plausibly work now :) :)

...but I'm still going to ask for a test.

File: pkg/apiserver/api_installer.go (hunk 14 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: Done.

File: pkg/apiserver/api_installer.go (hunk 13 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: Done.

File: pkg/apiserver/apiserver.go (hunk 10 above)
Comment (madhusudancs) [inline pkg/apiserver/apiserver.go]: Fixed.

File: pkg/genericapiserver/genericapiserver.go (hunk 11 above)
Comment (madhusudancs) [inline pkg/genericapiserver/genericapiserver.go]: Fixed.

**Comment by madhusudancs** [comment]
> ...but I'm still going to ask for a test.

Figuring out how to write a test for this :) Will push the test as soon as I have it working.

File: pkg/apiserver/api_installer.go (hunk 6 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: Done.

File: pkg/apiserver/api_installer.go (hunk 9 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: Fixed with examples. I think that's the best way to make this clear.

File: pkg/apiserver/apiserver.go (hunk 12 above)
Comment (madhusudancs) [inline pkg/apiserver/apiserver.go]: As we discussed offline, this is required to plumb the value from master to api_installer.

**Comment by madhusudancs** [comment]
@lavalamp Added a test and addressed other comments. I have created new commits (last two) for all the changes since last review to make the reviews easier. PTAL.

**Comment by madhusudancs** [comment]
@k8s-bot test this issue: #22088

File: pkg/apiserver/apiserver_test.go [hunk 15]
```diff
@@ -144,6 +152,11 @@ func addTestTypes() {
 	api.Scheme.AddKnownTypes(testInternalGroupVersion,
 		&apiservertesting.Simple{}, &apiservertesting.SimpleList{}, &api.ListOptions{},
 		&apiservertesting.SimpleGetOptions{}, &apiservertesting.SimpleRoot{})
+	api.Scheme.AddKnownTypes(testGroup2Version,
+		&SimpleXGSubresource{})
+	api.Scheme.AddKnownTypes(testGroupVersion, &api.Pod{})
```
Comment (lavalamp) [inline pkg/apiserver/apiserver_test.go]: Is registering pod necessary?

File: pkg/apiserver/api_installer.go [hunk 16]
```diff
@@ -110,97 +110,94 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
...
+// restMapping returns rest mapper for the resource, or the resource's parent if it is a sub resource.
+// Example REST paths that this mapper maps.
+// 1. Resource only, no subresource:
+//      Resource:    batch/v1.Job
+//      REST path:   /apis/batch/v1/namespaces/{namespace}/job/{name}
+// 2. Subresource and its parent belong to different API groups and/or versions:
+//      Resource:    extensions/v1beta1.ReplicaSet
+//      Subresource: autoscaling/v1.Scale
+//      REST path:   /apis/extensions/v1beta1/namespaces/{namespace}/replicaset/{name}/scale
+func (a *APIInstaller) restMapping(resource string) (*meta.RESTMapping, error) {
```
Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: 'resource' should never be a subresource, correct? Please fix the comments :)

File: pkg/apiserver/apiserver_test.go [hunk 17]
```diff
@@ -3157,6 +3170,106 @@ func TestUpdateChecksAPIVersion(t *testing.T) {
...
+	}
+	var itemOut SimpleXGSubresource
+	body, err := extractBody(resp, &itemOut)
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	if itemOut.Name != subresourceStorage.item.Name {
+		t.Errorf("Unexpected data: %#v, expected %#v (%s)", itemOut, subresourceStorage.item, string(body))
+	}
```
Comment (lavalamp) [inline pkg/apiserver/apiserver_test.go]: Test that the returned item is the correct group/version/kind.

File: pkg/apiserver/apiserver_test.go [hunk 18]
```diff
@@ -144,6 +152,11 @@ func addTestTypes() {
 	api.Scheme.AddKnownTypes(testInternalGroupVersion,
 		&apiservertesting.Simple{}, &apiservertesting.SimpleList{}, &api.ListOptions{},
 		&apiservertesting.SimpleGetOptions{}, &apiservertesting.SimpleRoot{})
+	api.Scheme.AddKnownTypes(testGroup2Version,
+		&SimpleXGSubresource{})
+	api.Scheme.AddKnownTypes(testGroupVersion, &api.Pod{})
+	api.Scheme.AddKnownTypes(testInternalGroup2Version,
+		&SimpleXGSubresource{})
```
Comment (lavalamp) [inline pkg/apiserver/apiserver_test.go]: Maybe register this in the other group, too, to prove that the correct one is chosen.

File: pkg/apiserver/api_installer.go (hunk 16 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: Done.

I removed ", or the resource's parent if it is a sub resource" part from the first line. But I am leaving the subresource example as is. I think that adds some clarity about how the subresource paths are mapped.

File: pkg/apiserver/apiserver_test.go (hunk 15 above)
Comment (madhusudancs) [inline pkg/apiserver/apiserver_test.go]: Removed.

Sorry, that was oversight on my part.

File: pkg/apiserver/apiserver_test.go (hunk 18 above)
Comment (madhusudancs) [inline pkg/apiserver/apiserver_test.go]: Done.

File: pkg/apiserver/api_installer.go [hunk 19]
```diff
@@ -110,97 +110,94 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
...
 		}
 	}
+	if fqKindToRegister.IsEmpty() {
+		return unversioned.GroupVersionKind{}, fmt.Errorf("unable to locate fully qualified kind for %v: found %v when registering for %v", reflect.TypeOf(object), fqKinds, a.group.GroupVersion)
+	}
+	return fqKindToRegister, nil
+}
 
-	kind := fqKindToRegister.Kind
+// restMapping returns rest mapper for the resource, or the resource's parent if it is a sub resource.
```
Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: Remove second half of sentence and replace with something saying not to pass a subresource to this function.

Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: Oh, I guess you haven't pushed yet :) sorry!

File: pkg/apiserver/api_installer.go (hunk 16 above)
Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: I agree the example is helpful.

Comment (liggitt) [inline pkg/apiserver/api_installer.go]: can you clarify the comment... it looks like you could pass in "batch/v1.Job" as the `resource` arg

**Comment by liggitt** [comment]
do we actually want to add "replicationcontrollers/scale" in v1 to prove this works?

**Comment by deads2k** [comment]
> do we actually want to add "replicationcontrollers/scale" in v1 to prove this works?

Great idea.

**Comment by lavalamp** [comment]
That may be a good idea.

**Comment by madhusudancs** [comment]
@liggitt @deads2k 

> do we actually want to add "replicationcontrollers/scale" in v1 to prove this works? I assume we would make it use the autoscaling/v1 Scale kind.

Yes. That was how I started with this whole cross group registration thing, that's how I ran into all sorts of problems that I am fixing right now ;-) There is a PR for that already - #21910

File: pkg/apiserver/apiserver_test.go (hunk 17 above)
Comment (madhusudancs) [inline pkg/apiserver/apiserver_test.go]: Done.

**Comment by madhusudancs** [comment]
@lavalamp I have pushed the changes. PTAL.

File: pkg/apiserver/apiserver_test.go [hunk 20]
```diff
@@ -3157,6 +3172,121 @@ func TestUpdateChecksAPIVersion(t *testing.T) {
...
+	}
+
+	// Test if the returned object has the expected group, version and kind
+	// We are directly unmarshaling JSON here because TypeMeta cannot be decoded through the
+	// installed decoders. TypeMeta cannot be decoded because it is added to the ignored
+	// conversion type list in API scheme and hence cannot be converted from input type object
+	// to output type object. So it's values don't appear in the decoded output object.
+	decoder := json.NewDecoder(strings.NewReader(body))
+	var itemFromBody SimpleXGSubresource
+	decoder.Decode(&itemFromBody)
```
Comment (lavalamp) [inline pkg/apiserver/apiserver_test.go]: check error return!

File: pkg/apiserver/api_installer.go (hunk 16 above)
Comment (lavalamp) [inline pkg/apiserver/api_installer.go]: +1, clarify what resource accepts.

**Comment by lavalamp** [comment]
Two nits, then LGTM

File: pkg/apiserver/apiserver_test.go [hunk 21]
```diff
@@ -3157,6 +3172,121 @@ func TestUpdateChecksAPIVersion(t *testing.T) {
...
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	// Test if the returned object has the expected group, version and kind
+	// We are directly unmarshaling JSON here because TypeMeta cannot be decoded through the
+	// installed decoders. TypeMeta cannot be decoded because it is added to the ignored
+	// conversion type list in API scheme and hence cannot be converted from input type object
+	// to output type object. So it's values don't appear in the decoded output object.
+	decoder := json.NewDecoder(strings.NewReader(body))
```
Comment (liggitt) [inline pkg/apiserver/apiserver_test.go]: This doesn't seem right… we can't decode an API object using a real codec?

Comment (madhusudancs) [inline pkg/apiserver/apiserver_test.go]: @liggitt We can decode a serialized API object into a destination, not just `TypeMeta`. In fact we can decode `TypeMeta` just fine, but the decoded values won't be copied to the destination object (`itemOut` here) and that is by design. It would be dangerous to overwrite `TypeMeta`. We could probably implement a converter to allow overwriting `TypeMeta` in tests, but I don't think it is worth spending time on it.

I think this is a reasonable workaround. Do you have any other alternative suggestions?

File: pkg/apiserver/api_installer.go (hunk 16 above)
Comment (madhusudancs) [inline pkg/apiserver/api_installer.go]: Done.

File: pkg/apiserver/apiserver_test.go (hunk 20 above)
Comment (madhusudancs) [inline pkg/apiserver/apiserver_test.go]: Done.

**Comment by madhusudancs** [comment]
@lavalamp Addressed the comments. PTAL.

File: pkg/apiserver/apiserver_test.go (hunk 21 above)
Comment (lavalamp) [inline pkg/apiserver/apiserver_test.go]: @liggett it's that I asked for the version/kind to be checked, and our codec consumes that information and leaves those fields blank.

**Comment by lavalamp** [comment]
LGTM

**Comment by lavalamp** [comment]
Squash please, then apply the label :)

File: pkg/apiserver/apiserver_test.go (hunk 21 above)
Comment (liggitt) [inline pkg/apiserver/apiserver_test.go]: UnstructuredJSONScheme would do this, I think...
```
obj, gvk, err := runtime.UnstructuredJSONScheme.Decode(data, nil, nil)
// assert no err
// assert gvk matches
```

Comment (liggitt) [inline pkg/apiserver/apiserver_test.go]: I guess it's json decoding either way, I just thought there was an existing codepath we could use

Comment (madhusudancs) [inline pkg/apiserver/apiserver_test.go]: @liggitt Do you feel strongly about swapping the current unmarshaling code with this snippet?

Comment (liggitt) [inline pkg/apiserver/apiserver_test.go]: I guess not

Comment (madhusudancs) [inline pkg/apiserver/apiserver_test.go]: @liggitt Alright then, thanks!
